README.md

```diff
@@ -17,3 +17,329 @@ Implement a prototype job worker service that provides an API to run arbitrary L
+    //stderr(standard Error) of command/program executed
+    LogOutput ErrorOutput 
+    //if the command was finished it will be 0, otherwise if command successfully completes it will be 1
+    bool finished
+    //timestamp returns the starting time if just executed or finish time if finished
+    string timeStamp 
+}
+
+```
+#### Stop
+
+* The User should be able to stop the request based on the pid
+
+* When stopped the pid should be killed 
+
+* A response will be sent to client indicating process have been stopped along with the contents of the log
+
+```
+type StopRequest {
+    int pid
+}
+
+type StopResponse {
+    LogOutput Output
+    LogOutput errorOutput
+    bool isKilled
+}
+
+```
+
+#### Query 
+
+* The client should be able to check what processes are still running
+
+* The Client should be able fetch the logs and fetch the status of the process by pid and starting timestamp
+
+* The client should able to view list of all logs generated by the server
+
+* Client should be able to read the logs into memory and send it and will only grab up to a certain amout of bytes from
+the file to prevent maxing out string size or memory 
+
+
+* There will be only two types of Query
+
+    * QueryOneProcess:
+
+        * get information on one pid
+
+    * QueryRunningProcesses:
+
+        * get list of running processes from server
+
+    * QueryListOfLogs
+
+        * get list of all logs generated by server 
+
+```
+type ProcessInfo {
+    int32 pid 
+    string startingTimeStamp 
+    string processName 
+    bool isRunning 
+    string logPath 
+}
+
+type QueryPidRequest {
+    int pid 
+    string startingTimeStamp
+}
+
+type QueryPidResponse {
+    LogOutput output 
+    LogOutput errorOutput 
+    bool isRunning 
+}
+
+type QueryRunningProcessesRequest {
+    //will be empty since the server just needs to verify its a QueryRunningProcessesRequest
+}
+    
+type QueryRunningProcessesResponse {
+     ProcessInfo[] processTable = 1;
+}
+
+type QueryListOfLogsRequest {
+    //will be empty since the server just needs to verify its a QueryListOfLogsRequest
+}
+
+type QueryListOfLogsResponse {
+    repeated ProcessInfo processTable = 1;
+}
+
+type QueryRunningProcessesResponse {
+     ProcessInfo[] processTable = 1;
+}
+type QueryRequest {
+    variableField queryCommand {
+        QueryPidRequest queryPid 
+        QueryRunningProcessesRequest queryRunningProcesses 
+        QueryListOfLogsResponse responseFromListOfLogs
+    }
+}
+
+type QueryResponse {
+variableField responseFromQuery {
+        QueryPidResponse responseFromQueryPid
+        QueryRunningProcessesResponse responseFromRunningProcesses
+        QueryListOfLogsResponse responseFromListOfLogs 
+    }
+}
+
+```
+
+
+### Implementation Overview 
+
+* Using GRPC so the above methods and types with be generated via protocol buffers (will be using libprotoc 3.11.4)
+
+####  DataStore
+
+* we can use sync.Map in Go to implement a set of structs to store process info
```

Comment:
Use a regular map with a mutex instead, per `sync.Map` docs:

> The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content.

Reply:
noted, I'll revisit, and think more about the DataStore design